sycl/test-e2e/InvokeSimd/Regression/nbarrier_multiple_wg.cpp

```diff
@@ -0,0 +1,214 @@
+// TODO: enable on Windows once driver is ready
+// NOTE: named barrier supported only since PVC
+// REQUIRES: gpu-intel-pvc && linux
+// UNSUPPORTED: cuda || hip
+//
+// TODO: enable when Jira issue resolved, currently fail with VISALTO enable
+// XFAIL: gpu-intel-pvc
+//
+// RUN: %{build} -fno-sycl-device-code-split-esimd -Xclang -fsycl-allow-func-ptr -o %t.out
+// RUN: env IGC_VCSaveStackCallLinkage=1 IGC_VCDirectCallsOnly=1 %{run} %t.out
+//
+// VISALTO enable run
+// RUN: env IGC_VISALTO=63 IGC_VCSaveStackCallLinkage=1 IGC_VCDirectCallsOnly=1 %{run} %t.out
+
+/*
+ * Test checks support of named barrier in invoke_simd context with multiple
+ * ESIMD work-groups. Producers store to SLM; consumers read SLM and store
+ * data to surface.
+ */
+
+#include <sycl/ext/intel/esimd.hpp>
+#include <sycl/ext/oneapi/experimental/invoke_simd.hpp>
+#include <sycl/sycl.hpp>
+
+#include <sycl/ext/intel/esimd.hpp>
+#include <sycl/ext/intel/experimental/esimd/memory.hpp>
+#include <sycl/ext/oneapi/experimental/invoke_simd.hpp>
+#include <sycl/sycl.hpp>
+
+#include <functional>
+#include <iostream>
+#include <type_traits>
+
+// TODO: When gpu driver can pass/accept accessor by value,
+// the work-around undef #ifdef US_ACC_PTR should be removed.
+#define USE_ACC_PTR
+
+using namespace sycl;
+using namespace sycl::ext::oneapi::experimental;
+namespace esimd = sycl::ext::intel::esimd;
+namespace experimental_esimd = sycl::ext::intel::experimental::esimd;
+
+constexpr int VL = 16;
+template <int case_num> class KernelID;
+
+template <unsigned Groups, unsigned Threads, unsigned Size>
+ESIMD_INLINE void ESIMD_CALLEE_nbarrier(local_accessor<int, 1> LocalAcc,
+                                        int groupID, int localID,
+                                        int *o) SYCL_ESIMD_FUNCTION {
+  // 1 named barrier, id 0 reserved for unnamed
+  constexpr unsigned bnum = 2;
+  constexpr unsigned bid = 1;
+  experimental_esimd::named_barrier_init<bnum>();
+
+  unsigned int group_off = VL * groupID * Threads;
+  unsigned int globalID = groupID * Threads + localID;
+  unsigned int global_off = VL * globalID;
+
+  esimd::simd<int, VL> val(localID);
+
+  constexpr unsigned producers = Threads / 2;
+  constexpr unsigned consumers = Threads / 2;
+
+  // thread with even local id is producer in each work-group
+  bool is_producer = localID % 2 == 0;
+  bool is_consumer = !is_producer;
+  // only-producer or only-comsumer modes
+  unsigned int flag = is_producer ? 0x1 : 0x2;
+
+  // producer writes to SLM, consumer reads what producer wrote
+  unsigned int off_slm =
+      static_cast<uint32_t>(
+          reinterpret_cast<std::uintptr_t>(LocalAcc.get_pointer())) +
+      (is_producer ? global_off : (global_off - VL)) * sizeof(int);
```

Comment:
is the difference between producer and consumer so we intentionally get overlapping regions? if so we can add a comment?